main.py

```diff
@@ -0,0 +1,848 @@
+        image = self._load_image(index)
+        target = self._load_target(index)
+        lang = self._load_lang(index)
+        lang = self.map_lang_code[lang]
+
+        if self.transforms is not None:
+            image, target = self.transforms(image, target)
+
+        return image, target, lang
+
+    def __len__(self) -> int:
+        return len(self.captions)
+
+class TrainState(train_state.TrainState):
+    dropout_rng: jnp.ndarray
+
+    def replicate(self):
+        return jax_utils.replicate(self).replace(dropout_rng=shard_prng_key(self.dropout_rng))
+
+
+
+
+def write_train_metric(summary_writer, train_metrics, train_time, step):
+    summary_writer.scalar("train_time", train_time, step)
+
+    train_metrics = get_metrics(train_metrics)
+    for key, vals in train_metrics.items():
+        tag = f"train/{key}"
+        for i, val in enumerate(vals):
+            summary_writer.scalar(tag, val, step - len(vals) + i + 1)
+
+
+def write_eval_metric(summary_writer, eval_metrics, step):
+    writable_eval_metrics = {}
+    for key,value in eval_metrics.items():
+        if isinstance(value,dict):
+            for sub_key,sub_value in value.items():
+                writable_eval_metrics[sub_key+"/"+key] = sub_value
+        else:
+            writable_eval_metrics[key]=value
+
+
+    for metric_name, value in writable_eval_metrics.items():
+        if metric_name =="loss":
+            summary_writer.scalar(f"eval_{metric_name}", value, step)
+        else:
+            summary_writer.scalar(f"{metric_name}", value, step)    
+
+
+def create_learning_rate_fn(
+    train_ds_size: int, train_batch_size: int, num_train_epochs: int, num_warmup_steps: int, learning_rate: float
+) -> Callable[[int], jnp.array]:
+    """Returns a linear warmup, linear_decay learning rate function."""
+    steps_per_epoch = train_ds_size // train_batch_size
+    num_train_steps = steps_per_epoch * num_train_epochs
+    warmup_fn = optax.linear_schedule(init_value=0.0, end_value=learning_rate, transition_steps=num_warmup_steps)
+    decay_fn = optax.linear_schedule(
+        init_value=learning_rate, end_value=0, transition_steps=num_train_steps - num_warmup_steps
+    )
+    schedule_fn = optax.join_schedules(schedules=[warmup_fn, decay_fn], boundaries=[num_warmup_steps])
+    return schedule_fn
+
+# utils
+def mb_item(x):
+    return x.item() if hasattr(x, "item") else x
+
+#checkpoint functions
+def save_model_checkpoint(model, save_dir, state, logger, organization,  with_opt:bool=False, push_to_hub:bool=False, overwrite=False, **kwargs):
+    state = jax_utils.unreplicate(state)
+    logger.info(f"Saving Checkpoint in {save_dir}")
+    ckpt_save_dir = f"{save_dir}/ckpt-{mb_item(state.step)-1}"
+    if os.path.exists(ckpt_save_dir) and not overwrite:
+        logger.info("checkpoint exists, skipping overwrite")
+    else:
+        model.save_pretrained(
+            ckpt_save_dir,
+            params=state.params,
+            push_to_hub=False,
```

Comment:
think you can just do `push_to_hub=push_to_hub` here instead of the special "push_to_hub" code below